util/bitmap/concurrent.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bitmap
+
+import (
+	"sync/atomic"
+	"unsafe"
+)
+
+const (
+	segmentWidth             = 32
```

Comment:
Why not use 8 for segmentWidth? Will 8 reduce contention when concurrency is large?

Reply:
Because the atomic type is at least 32bit, like atomic_int.

Reply:
Can we use a lock for every element instead of CAS?

Reply:
Locking needs additionaly memory space, and it is not so convenient as CAS.

Reply:
1. The memory space is small.
2. If we use `lock`, we can use fine-grain lock which uses only 8 bits instead 32 bits now. It can decrease the possibility of concurrency contention.

Reply:
1. The lock structure in go (`sync.Mutex`) uses 64bits, it will occupy 8GB memory space if there are 1G (2^30) tuples, but 1 GB when using this `ConcurrentBitmap`.
2. The possibility of concurrency contention is small under this case so that it cannot be the performance bottleneck.

Reply:
I do not mean to use a lock for each row, I mean use a lock for every byte which holds 8 rows.

Reply:
> The possibility of concurrency contention is small under this case

How do we know this?

Reply:
It is really related to data, but I believe it is small in normal cases. It's overhead is little in comparison to the general join operator.  If we find it introduces a lot of overhead in some special cases in the future, we can optimize it on that time.